### ISSUE TITLE
Error: /tmp/1.m3u8 does not exists

### ISSUE THREAD

**Issue by ronovar** [yuanchuan/node-watch#49]
I im having ffmpeg that is generating upon node start in tmp folder m3u8 files that holds indexes for ts segment files..so i need to watch m3u8 file to file change and then extract indexes from that file.

node-watch works excellent in terms of low cpu usage and instantly it get triggered from OS when m3u8 file changes...that works flawlessly.

But problem is when i start program ffmpeg needs sometime 2seconds to generate m3u8 file sometimes 10seconds...so time creation of m3u8 file is random.

And when i call on start node.js application and run watch for file 1.m3u8 i get this error:

 Error: /tmp//function (req, res, next) {
    app.handle(req, res, next);
  }_.m3u8 does not exist.

So i need if you could rewrite code in watch.js line 371 to return to watch function and to call it again in for example one second...when file will be generated...or if you have better idea you can add it to this line...

here is original code that throws error when file not exists...maybe adding check if file exists and then settimeout to call function again?

if (!is.exists(fpath)) {
    watcher.emit('error',
      new Error(fpath + ' does not exist.')
    );
  }

So in short description problem is that watcher exits with error when file does not exists...if you could fix this to if file not exists call watcher again in one second...if it exists continuo to execute code like now

**Comment by yuanchuan** [comment]
You don't need to watch for files individually.  How about to setup a watcher with filtering on the parent tmp folder?

**Comment by ronovar** [comment]
Yes you are right...i forget for this (yesterday i was reading your api examples...and i see that this is the way...watching for m3u8 with filter and when exists do other stuffs..like reading file and extracting contents from file)

Thanks...for great answer.

**Comment by yuanchuan** [comment]
Glad to help.

**Comment by ronovar** [comment]
Thanks..i have one more question...if i run let's say 100 same watchers on same file it will be cpu usage high or is this ok to run?

I have this defined code to start watching when user request file:

stream_play[user] = watch('/tmp/'+stream+'_.m3u8', function(event, name) {
    /* Here i im doing res.write on every change on m3u8 file when watcher is triggered */
})

so after response is end i do this:

res.on('close', function(){
	/* CLOSE - m3u8 watcher */
	stream_play[user].close();
	/* CLOSE - stream */
	res.end();
})

So i need to start for each user watcher so that i can do fs.readFileSync on segment file to write to user...this works great and cpu usage is low ( i tested on 3users)..but if 100users or more request same file it will watch for stream_play[1...100] same watcher...so i im asking if this is good solution or do you have another how to use?

**Comment by yuanchuan** [comment]
I haven't tested myself but it definitely will take more server resources as the numbers of watchers increase. 

I think it's not scalable that way according to your descriptions above. You could try to separate the whole program into at least two parts ([pub-sub](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) pattern):

1. Watch for file changes and broadcast.  ( Watch a parent directory is enough)
2. Handle user requests and subscribe for changes.

**Comment by ronovar** [comment]
Thanks i was thinking about this that stream_play[user] will hold id of stream and if this stream exists(file watched) to return buffer of this stream something linke this:

if (stream[id] = id_requested) {
   res.write(stream_play[id]));
} else {
      stream_play[id] = watch('/tmp/'+stream+'_.m3u8', function(event, name) {
         stream[id] = readfilebuffer('/tmp/'+file_ts);
         return stream[id];
    })}
So the idea is to return pointer to existing watched file and to read for other 100 clients requesting same id stream from memory....but i im not sure if is possible with node...this is like pointer to id of stream that is read from watcher to memory. but res.write(stream_play[id])); is not refreeshed anymore because it reads only once variable stream_play[id])...and i would not want to use setinterval because this will be cpu insufficient.

So any idea how to use this but in short quesrtion is how to return from watcher buffer of read chunk file and then refresh that value so that one watch per stream id...this way i don't need to watch 100 times same id just to point other client request to watcher id that is in memory and that is refreshed buffer from one watch of this id in memory..

i think you got the idea what i would like to describe

**Comment by crh3675** [comment]
So you want to write buffers as they are available to all connected clients? 

You really want to be piping data to a socket that all clients can communicate with, not base your solution on file watching. SocketIO can do this very easily if you have the clients subscribe to a channel that receives the buffers.  Note that the buffers will be incremental and no cumulative as you don't want to be sending via pub/sub 20MB of data.  I use Redis as a memory-cache and use SockerIO to connect to web clients.

You would publish to a channel with a message like:
`publish video-xyz nextBuffer`

SocketIo would subscribe once in your application but distribute to all connected clients. It sounds to me that your problem involves more technology than you have laid out.

**Comment by ronovar** [comment]
Thanks...i got your idea but in node js i don't know how to point each client that request stream id same to stream it...like on starting node js i put stream_start[id_of_stream] = mycodethatisreadinsidewatcherm3u8file; and later in authorize function i need to point to stream_start[id_of_stream] and stream is started but it is not refreshed from res.write so if i call it again it is refreshed but this is cpu intensive.
